components/zcash_note_encryption/src/lib.rs

```diff
@@ -1,8 +1,464 @@
+///
+/// ```
+/// extern crate ff;
+/// extern crate rand_core;
+/// extern crate zcash_primitives;
+///
+/// use ff::Field;
+/// use rand_core::OsRng;
+/// use zcash_primitives::{
+///     consensus::TestNetwork,
+///     memo::MemoBytes,
+///     sapling::{
+///         keys::{OutgoingViewingKey, prf_expand},
+///         note_encryption::{sapling_note_encryption},
+///         Diversifier, PaymentAddress, Rseed, ValueCommitment
+///     },
+/// };
+///
+/// let mut rng = OsRng;
+///
+/// let diversifier = Diversifier([0; 11]);
+/// let pk_d = diversifier.g_d().unwrap();
+/// let to = PaymentAddress::from_parts(diversifier, pk_d).unwrap();
+/// let ovk = Some(OutgoingViewingKey([0; 32]));
+///
+/// let value = 1000;
+/// let rcv = jubjub::Fr::random(&mut rng);
+/// let cv = ValueCommitment {
+///     value,
+///     randomness: rcv.clone(),
+/// };
+/// let rcm = jubjub::Fr::random(&mut rng);
+/// let note = to.create_note(value, Rseed::BeforeZip212(rcm)).unwrap();
+/// let cmu = note.cmu();
+///
+/// let mut enc = sapling_note_encryption::<_, TestNetwork>(ovk, note, to, MemoBytes::empty(), &mut rng);
+/// let encCiphertext = enc.encrypt_note_plaintext();
+/// let outCiphertext = enc.encrypt_outgoing_plaintext(&cv.commitment().into(), &cmu, &mut rng);
+/// ```
+pub struct NoteEncryption<D: Domain> {
+    epk: D::EphemeralPublicKey,
+    esk: D::EphemeralSecretKey,
+    note: D::Note,
+    to: D::Recipient,
+    memo: D::Memo,
+    /// `None` represents the `ovk = ⊥` case.
+    ovk: Option<D::OutgoingViewingKey>,
+}
+
+impl<D: Domain> NoteEncryption<D> {
+    /// Construct a new note encryption context for the specified note,
+    /// recipient, and memo.
+    pub fn new(
+        ovk: Option<D::OutgoingViewingKey>,
+        note: D::Note,
+        to: D::Recipient,
+        memo: D::Memo,
+    ) -> Self {
+        let esk = D::derive_esk(&note).expect("ZIP 212 is active.");
+        Self::new_with_esk(esk, ovk, note, to, memo)
+    }
+
+    /// For use only with Sapling.
+    pub fn new_with_esk(
+        esk: D::EphemeralSecretKey,
+        ovk: Option<D::OutgoingViewingKey>,
+        note: D::Note,
+        to: D::Recipient,
+        memo: D::Memo,
+    ) -> Self {
+        NoteEncryption {
+            epk: D::ka_derive_public(&note, &esk),
+            esk,
+            note,
+            to,
+            memo,
+            ovk,
+        }
+    }
+
+    /// Exposes the ephemeral secret key being used to encrypt this note.
+    pub fn esk(&self) -> &D::EphemeralSecretKey {
+        &self.esk
+    }
+
+    /// Exposes the ephemeral public key being used to encrypt this note.
+    pub fn epk(&self) -> &D::EphemeralPublicKey {
+        &self.epk
+    }
+
+    /// Generates `encCiphertext` for this note.
+    pub fn encrypt_note_plaintext(&self) -> [u8; ENC_CIPHERTEXT_SIZE] {
+        let pk_d = D::get_pk_d(&self.note);
+        let shared_secret = D::ka_agree_enc(&self.esk, &pk_d);
+        let key = D::kdf(shared_secret, &self.epk);
```

Comment:
The KDF takes the encoding `ephemeralKey` as input, not the abstract point `epk`.